### PR TITLE
Benchmark tool for imgnet

### PR DESCRIPTION
**Pull request by typhoonzero**

### CODE REVIEW

**Comment by typhoonzero** [comment]
Added `scope` paramter in `ParallelExecutor`, when train with NCCL2 distributed mode, the default scope will have an initialized NCCLID variable which should not be used for testing with `ParalllelExecutor`

File: benchmark/fluid/Dockerfile [hunk 1]
```diff
@@ -11,6 +11,9 @@ RUN ln -s /usr/lib/x86_64-linux-gnu/libcudnn.so.7 /usr/lib/libcudnn.so && ln -s
 # Add "ENV http_proxy=http://ip:port" if your download is slow, and don't forget to unset it at runtime.
 # exmaple: unset http_proxy && unset https_proxy && python fluid_benchmark.py ...
 
+ENV http_proxy=http://172.19.32.197:9988/
```
Comment (gongweibao) [inline benchmark/fluid/Dockerfile]: It doesn't work now.

File: benchmark/fluid/fluid_benchmark.py [hunk 2]
```diff
@@ -326,44 +311,53 @@ def main():
...
-        else:
-            train(*train_args)
-    else:
-        if args.device == "CPU":
-            raise Exception("Only support GPU perf with parallel exe")
-        train_args.extend([nccl_id_var, num_trainers, trainer_id])
-        train_parallel(*train_args)
+        nccl_id_var, num_trainers, trainer_id = append_nccl2_prepare(
+            trainer_id, startup_prog)
+    # if args.gpus == 1:
```
Comment (gongweibao) [inline benchmark/fluid/fluid_benchmark.py]: Comment?

File: paddle/fluid/API.spec [hunk 3]
```diff
@@ -74,7 +74,7 @@ paddle.fluid.InferenceTranspiler.transpile ArgSpec(args=['self', 'program', 'pla
 paddle.fluid.memory_optimize ArgSpec(args=['input_program', 'skip_opt_set', 'print_log', 'level'], varargs=None, keywords=None, defaults=(None, False, 0))
 paddle.fluid.release_memory ArgSpec(args=['input_program', 'skip_opt_set'], varargs=None, keywords=None, defaults=(None,))
 paddle.fluid.DistributeTranspilerConfig.__init__ 
-paddle.fluid.ParallelExecutor.__init__ ArgSpec(args=['self', 'use_cuda', 'loss_name', 'main_program', 'share_vars_from', 'exec_strategy', 'build_strategy', 'num_trainers', 'trainer_id'], varargs=None, keywords='kwargs', defaults=(None, None, None, None, None, 1, 0))
+paddle.fluid.ParallelExecutor.__init__ ArgSpec(args=['self', 'use_cuda', 'loss_name', 'main_program', 'share_vars_from', 'exec_strategy', 'build_strategy', 'num_trainers', 'trainer_id', 'scope'], varargs=None, keywords='kwargs', defaults=(None, None, None, None, None, 1, 0, None))
```
Comment (gongweibao) [inline paddle/fluid/API.spec:68]: Maybe split to another PR can speed the merge process.

**Review by gongweibao — CHANGES_REQUESTED** [review]

File: benchmark/fluid/Dockerfile [hunk 4]
```diff
@@ -23,9 +26,14 @@ ADD https://raw.githubusercontent.com/PaddlePaddle/cloud/develop/docker/paddle_k
 ADD https://raw.githubusercontent.com/PaddlePaddle/cloud/develop/docker/k8s_tools.py /root
 RUN chmod +x /usr/bin/paddle_k8s
 
+ENV http_proxy=http://172.19.57.47:8899/
```
Comment (gongweibao) [inline benchmark/fluid/Dockerfile]: Do we need to reserve them?

**Review by reyoung — DISMISSED** [review]

**Review by gongweibao — DISMISSED** [review]
LGTM

File: paddle/fluid/API.spec [hunk 5]
```diff
@@ -65,7 +65,7 @@ paddle.fluid.InferenceTranspiler.transpile ArgSpec(args=['self', 'program', 'pla
...
```
Comment (panyx0718) [inline paddle/fluid/API.spec]: where is the new ParallelExecutor used?

File: python/paddle/fluid/parallel_executor.py [hunk 6]
```diff
@@ -73,6 +73,7 @@ def __init__(self,
                  build_strategy=None,
                  num_trainers=1,
                  trainer_id=0,
+                 scope=None,
```
Comment (panyx0718) [inline python/paddle/fluid/parallel_executor.py:77]: comment?

**Comment by panyx0718** [comment]
Does it affect testing if the ncclid is in the scope?

**Comment by typhoonzero** [comment]
@panyx0718 NCCLID in the scope will affect testing using ParallelExecutor, pass a new scope and share `local_scopes` to the test executor will be fine.

**Review by reyoung — APPROVED** [review]

**Review by panyx0718 — APPROVED** [review]

**Pull request merged**